RFC-0026-logging-system.md

```diff
@@ -0,0 +1,481 @@
+| `TORCH_CHECK_INDEX(cond, ...)`           | `c10::IndexError`              |
+| `TORCH_CHECK_VALUE(cond, ...)`           | `c10::ValueError`              |
+| `TORCH_CHECK_TYPE(cond, ...)`            | `c10::TypeError`               |
+| `TORCH_CHECK_NOT_IMPLEMENTED(cond, ...)` | `c10::NotImplementedError`     |
+
+There is some documentation on error macros [here](https://github.com/pytorch/pytorch/blob/72e4aab74b927c1ba5c3963cb17b4c0dce6e56bf/c10/util/Exception.h#L344-L362)
+
+The reason why C++ preprocessor macros are used, rather than function calls, is
+to ensure that the compiler can optimize for the `cond == true` branch. In
+other words, if an error does not get raised, overhead is minimized.
+
+### C++ error classes
+
+The primary error class in C++ is `c10::Error`. Documentation and declaration
+are
+[here](https://github.com/pytorch/pytorch/blob/72e4aab74b927c1ba5c3963cb17b4c0dce6e56bf/c10/util/Exception.h#L21-L28).
+`c10::Error` is a subclass of `std::exception`.
+
+There are other error classes which are child classes of `c10::Error`, defined
+[here](https://github.com/pytorch/pytorch/blob/72e4aab74b927c1ba5c3963cb17b4c0dce6e56bf/c10/util/Exception.h#L195-L236).
+
+When these errors propagate to Python, they are each converted to a different
+Python error class:
+
+| C++ error class                 | Python error class         |
+| ------------------------------- | -------------------------- |
+| `c10::Error`                    | `RuntimeError`             |
+| `c10::IndexError`               | `IndexError`               |
+| `c10::ValueError`               | `ValueError`               |
+| `c10::TypeError`                | `TypeError`                |
+| `c10::NotImplementedError`      | `NotImplementedError`      |
+| `c10::EnforceFiniteError`       | `ExitException`            |
+| `c10::OnnxfiBackendSystemError` | `ExitException`            |
+| `c10::LinAlgError`              | `torch.linalg.LinAlgError` |
+
+
+## PyTorch C++ Warnings
+
+When warnings propagate from C++ to Python, they are converted to a Python
+`UserWarning`. Whatever is in `...` will get concatenated into the warning
+message using `operator<<`.
+
+* `TORCH_WARN(...)`
+  - [Definition](https://github.com/pytorch/pytorch/blob/72e4aab74b927c1ba5c3963cb17b4c0dce6e56bf/c10/util/Exception.h#L515-L530)
+
+* `TORCH_WARN_ONCE(...)`
+  - [Definition](https://github.com/pytorch/pytorch/blob/72e4aab74b927c1ba5c3963cb17b4c0dce6e56bf/c10/util/Exception.h#L557-L562)
+  - This macro only generates a warning the first time it is encountered during
+    run time.
+
+
+## Implementation details
+
+### C++ to Python Error Translation
+
+`c10::Error` and its subclasses are translated into their corresponding Python
+errors [in `CATCH_CORE_ERRORS`](https://github.com/pytorch/pytorch/blob/72e4aab74b927c1ba5c3963cb17b4c0dce6e56bf/torch/csrc/Exceptions.h#L54-L100).
+
+However, not all of the `c10::Error` subclasses in the table above appear here.
```

Comment:
Most likely oversight when these errors were added or when that macro was made :D